src/FEValues/cell_values.jl

```diff
@@ -34,137 +34,244 @@ utilizes scalar shape functions and `CellVectorValues` utilizes vectorial shape
+    dMdξ = fill(zero(SVector{rdim,T}) * T(NaN), n_geom_basefuncs, n_qpoints)
 
     for (qp, ξ) in enumerate(quad_rule.points)
         for i in 1:n_func_basefuncs
             dNdξ[i, qp], N[i, qp] = gradient(ξ -> value(func_interpol, i, ξ), ξ, :all)
         end
         for i in 1:n_geom_basefuncs
-            dMdξ[i, qp], M[i, qp] = gradient(ξ -> value(geom_interpol, i, ξ), ξ, :all)
+            dMdξ[i, qp], M[i, qp] = gradient(ξ -> value(geo_interpol, i, ξ), ξ, :all)
         end
     end
 
     detJdV = fill(T(NaN), n_qpoints)
 
-    CellScalarValues{dim,T,shape}(N, dNdx, dNdξ, detJdV, M, dMdξ, quad_rule, func_interpol, geom_interpol)
+    CellScalarValues{sdim,rdim,T,shape}(N, dNdx, dNdξ, detJdV, M, dMdξ, quad_rule, func_interpol, geo_interpol)
 end
 
 # CellVectorValues
-struct CellVectorValues{dim,T<:Real,refshape<:AbstractRefShape,M} <: CellValues{dim,T,refshape}
-    N::Matrix{Vec{dim,T}}
-    dNdx::Matrix{Tensor{2,dim,T,M}}
-    dNdξ::Matrix{Tensor{2,dim,T,M}}
+# TODO remove the assumption that all dimensions have to match.
+#   rdim = reference element dimension
+#   sdim = spatial dimension
+#   vdim = vector dimension (i.e. dimension of evaluation of what `value` should return)
+struct CellVectorValues{sdim,rdim,T<:Real,refshape<:AbstractRefShape,vdim} <: CellValues{sdim,rdim,T,refshape}
+    N::Matrix{SVector{vdim,T}} # vdim
+    dNdx::Matrix{SMatrix{vdim,sdim,T}} # vdim × sdim
+    dNdξ::Matrix{SMatrix{vdim,rdim,T}} # vdim × rdim
     detJdV::Vector{T}
     M::Matrix{T}
-    dMdξ::Matrix{Vec{dim,T}}
-    qr::QuadratureRule{dim,refshape,T}
+    dMdξ::Matrix{SVector{rdim,T}} # rdim
+    qr::QuadratureRule{rdim,refshape,T} #rdim
     # The following fields are deliberately abstract -- they are never used in
     # performance critical code, just stored here for convenience.
-    func_interp::Interpolation{dim,refshape}
-    geo_interp::Interpolation{dim,refshape}
+    func_interp::Interpolation{rdim,refshape} # rdim
+    geo_interp::Interpolation{rdim,refshape} # rdim
 end
 
-function CellVectorValues(quad_rule::QuadratureRule, func_interpol::Interpolation, geom_interpol::Interpolation=func_interpol)
-    CellVectorValues(Float64, quad_rule, func_interpol, geom_interpol)
+function CellVectorValues(quad_rule::QuadratureRule, func_interpol::Interpolation, geo_interpol::Interpolation=func_interpol; sdim::Int=getdim(geo_interpol), vdim::Int=getdim(func_interpol))
+    CellVectorValues(Float64, quad_rule, func_interpol, geo_interpol, sdim, vdim)
 end
 
-function CellVectorValues(::Type{T}, quad_rule::QuadratureRule{dim,shape}, func_interpol::Interpolation,
-        geom_interpol::Interpolation=func_interpol) where {dim,T,shape<:AbstractRefShape}
-
-    @assert getdim(func_interpol) == getdim(geom_interpol)
-    @assert getrefshape(func_interpol) == getrefshape(geom_interpol) == shape
+# FIXME sdim should be something like `length(value(geo_interpol))`
+# FIXME vdim should be something like `length(value(func_interpol))`
```

Comment:
Would this fix most of the constructor speed regression?

Reply:
If I can find a way to query `sdim` and `vdim` on compile time, likely yes. Currently I introduced the kwargs to pass the information manually (because it is not in the interpolations yet).

Reply:
Ah, right. We only know `rdim` from the interpolation! 
If a fast track is desired, maybe you could allow passing the kwargs as `Val`'s ?

Reply:
Well, I can't think of a scenario where the ctor will be the bottleneck (3µs is still fast enough, I think).

Reply:
Resolved it anyway. :)